lib/SILOptimizer/Transforms/TempRValueElimination.cpp

```diff
@@ -601,6 +602,28 @@ void TempRValueOptPass::tryOptimizeCopyIntoTemp(CopyAddrInst *copyInst) {
 
   LLVM_DEBUG(llvm::dbgs() << "  Success: replace temp" << *tempObj);
 
+  // If copyInst's source must be deinitialized, whether that must be done via
+  // a newly created destroy_addr.
+  //
+  // If lastLoadInst is a load or a copy_addr, then the deinitialization can be
+  // done in that instruction.
+  //
+  // This is necessary for correctness: otherwise, copies of move-only values
+  // would be introduced.
+  bool needToInsertDestroy = [&]() {
+    if (!needFinalDeinit)
+      return false;
+    if (lastLoadInst == copyInst)
+      return true;
+    if (auto *cai = dyn_cast<CopyAddrInst>(lastLoadInst)) {
+      return cai->getSrc() != tempObj || !cai->isTakeOfSrc();
+    }
+    if (auto *li = dyn_cast<LoadInst>(lastLoadInst)) {
+      return li->getOperand() != tempObj ||
+             li->getOwnershipQualifier() != LoadOwnershipQualifier::Take;
+    }
+    return true;
+  }();
```

Comment:
Maybe you can add an assert here to check if it's a move only type then `needToInsertDestroy` must be false